SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp

```diff
@@ -1463,6 +1465,175 @@ void USpatialNetDriver::ProcessRPC(AActor* Actor, UObject* SubObject, UFunction*
+			{
+				AActor* NetOwner = SpatialGDK::GetHierarchyRoot(Actor);
+
+				uint64 HierarchyAuthorityReceivedTimestamp = 0;
+
+				GetLatestAuthorityChangeFromHierarchy(NetOwner, HierarchyAuthorityReceivedTimestamp);
+
+				const float TimeSinceReceivingAuthInSeconds = double(FPlatformTime::Cycles64() - HierarchyAuthorityReceivedTimestamp) * FPlatformTime::GetSecondsPerCycle64();
+				const float MigrationBackoffTimeInSeconds = 1.0f;
+
+				if (TimeSinceReceivingAuthInSeconds < MigrationBackoffTimeInSeconds)
+				{
+					UE_LOG(LogSpatialOSNetDriver, Verbose, TEXT("Tried to change auth too early for actor %s"), *Actor->GetName());
+				}
+				else
+				{
+					const VirtualWorkerId NewAuthVirtualWorkerId = LoadBalanceStrategy->WhoShouldHaveAuthority(*Actor);
+					if (NewAuthVirtualWorkerId == SpatialConstants::INVALID_VIRTUAL_WORKER_ID)
+					{
+						UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Load Balancing Strategy returned invalid virtual worker for actor %s"), *Actor->GetName());
+					}
+					else
+					{
+						FMigrationInfo MigInfo = { Actor, NewAuthVirtualWorkerId };
+						OutMigrationInfo.Add(MigInfo);
+						CollectActorsToMigrate(NetOwner, Actor, NewAuthVirtualWorkerId, OutMigrationInfo, AdditionalObjectsToConsider);
+					}
+				}
+			}
+		}
+	}
+
+	// Processing actors that we pull from actor hierarchies, in order to migrate them together.
+	// This loop is extracted from UNetDriver::ServerReplicateActors_BuildConsiderList
+	for(auto ActorInfoIter = AdditionalObjectsToConsider.CreateIterator(); ActorInfoIter; ++ActorInfoIter)
+	{
+		FNetworkObjectInfo* ActorInfo = *ActorInfoIter;
+		AActor* Actor = ActorInfo->Actor;
+
+		if (Actor->IsPendingKillPending())
+		{
+			continue;
+		}
+
+		// Verify the actor is actually initialized (it might have been intentionally spawn deferred until a later frame)
+		if (!Actor->IsActorInitialized())
+		{
+			continue;
+		}
+
+		// Don't send actors that may still be streaming in or out
+		ULevel* Level = Actor->GetLevel();
+		if (Level->HasVisibilityChangeRequestPending() || Level->bIsAssociatingLevel)
+		{
+			continue;
+		}
+
+		if (Actor->NetDormancy == DORM_Initial && Actor->IsNetStartupActor())
+		{
+			continue;
+		}
+
+		ActorInfo->bPendingNetUpdate = false;
+
+		// Call PreReplication on all actors that will be considered
+		Actor->CallPreReplication(this);
+
+		// Add it to the consider list.
+		ConsiderList.Add(ActorInfo);
+	}
+}
+
+void USpatialNetDriver::ServerReplicateActors_ProcessMigration(const TArray<FMigrationInfo>& MigrationInfo)
+{
+	for(const FMigrationInfo& Info : MigrationInfo )
+	{
+		Sender->SendAuthorityIntentUpdate(*Info.ActorToMigrate, Info.Destination);
+
+		// If we're setting a different authority intent, preemptively changed to ROLE_SimulatedProxy
+		Info.ActorToMigrate->Role = ROLE_SimulatedProxy;
+		Info.ActorToMigrate->RemoteRole = ROLE_Authority;
+
+		Info.ActorToMigrate->OnAuthorityLost();
+	}
+}
```

Comment:
If we're refactoring logic - can we put all of this in its own class? NetDriver is a huge mess already

Reply:
It's not worth a class, since it has close to no reusability, but I'll put it in its own source file.

Reply:
+1 to trying to pull this out. Having two implementation files and a load of C style naming to indicate that it's a method of NetDriver but it really belongs to some other group is confusing.